src/arch/x86_64/CodeGen.zig

```diff
@@ -40777,980 +40426,1922 @@ fn floatCompilerRtAbiType(self: *CodeGen, ty: Type, other_ty: Type) Type {
+                try temp.toOffset(-off, cg);
+                while (try temp.toRegClass(true, .general_purpose, cg)) {}
+                try temp.toOffset(off, cg);
+                return true;
+            },
+        }
+    }
 
-fn promoteInt(self: *CodeGen, ty: Type) Type {
-    const pt = self.pt;
-    const zcu = pt.zcu;
-    const int_info: InternPool.Key.IntType = switch (ty.toIntern()) {
-        .bool_type => .{ .signedness = .unsigned, .bits = 1 },
-        else => if (ty.isAbiInt(zcu)) ty.intInfo(zcu) else return ty,
-    };
-    for ([_]Type{
-        .c_int,      .c_uint,
-        .c_long,     .c_ulong,
-        .c_longlong, .c_ulonglong,
-    }) |promote_ty| {
-        const promote_info = promote_ty.intInfo(zcu);
-        if (int_info.signedness == .signed and promote_info.signedness == .unsigned) continue;
-        if (int_info.bits + @intFromBool(int_info.signedness == .unsigned and
-            promote_info.signedness == .signed) <= promote_info.bits) return promote_ty;
+    fn toMemory(temp: *Temp, cg: *CodeGen) !bool {
+        const temp_tracking = temp.tracking(cg);
+        if (temp_tracking.short.isMemory()) return false;
+        const new_temp_index = cg.next_temp_index;
+        const ty = temp.typeOf(cg);
+        cg.temp_type[@intFromEnum(new_temp_index)] = ty;
+        const new_frame_index = try cg.allocFrameIndex(.initSpill(ty, cg.pt.zcu));
+        try cg.genSetMem(.{ .frame = new_frame_index }, 0, ty, temp_tracking.short, .{});
+        new_temp_index.tracking(cg).* = .init(.{ .load_frame = .{ .index = new_frame_index } });
+        try temp.die(cg);
+        cg.next_temp_index = @enumFromInt(@intFromEnum(new_temp_index) + 1);
+        temp.* = .{ .index = new_temp_index.toIndex() };
+        return true;
     }
-    return ty;
-}
 
-fn promoteVarArg(self: *CodeGen, ty: Type) Type {
-    if (!ty.isRuntimeFloat()) return self.promoteInt(ty);
-    switch (ty.floatBits(self.target.*)) {
-        32, 64 => return .f64,
-        else => |float_bits| {
-            assert(float_bits == self.target.cTypeBitSize(.longdouble));
-            return .c_longdouble;
-        },
+    // hack around linker relocation bugs
+    fn toBase(temp: *Temp, cg: *CodeGen) !bool {
+        const temp_tracking = temp.tracking(cg);
+        if (temp_tracking.short.isBase()) return false;
+        if (try temp.toMemory(cg)) return true;
+        const new_temp_index = cg.next_temp_index;
+        cg.temp_type[@intFromEnum(new_temp_index)] = temp.typeOf(cg);
+        const new_reg =
+            try cg.register_manager.allocReg(new_temp_index.toIndex(), abi.RegisterClass.gp);
+        try cg.genSetReg(new_reg, .usize, temp_tracking.short.address(), .{});
+        new_temp_index.tracking(cg).* = .init(.{ .indirect = .{ .reg = new_reg } });
+        try temp.die(cg);
+        cg.next_temp_index = @enumFromInt(@intFromEnum(new_temp_index) + 1);
+        temp.* = .{ .index = new_temp_index.toIndex() };
+        return true;
```

Comment:
is this hack stable enough to last until I fix said linker bugs? or going to introduce more instability into our CI?

Reply:
What the comment is referring to is that it should be valid to replace every use of `toBase` with `toMemory`, but that makes it possible to send a specific instruction to the linker (well `Emit.zig` specifically) that it mangles.  This function doesn't allow that form through for any instruction at all as a workaround, resulting in more instructions to be emitted.  It's not a hack in that it's unstable, just in that it should be reverted when the linker properly supports arbitrary x86_64 instructions, which is probably something I will have to implement.